### PR TITLE
[specific ci=Group6-VIC-Machine]Add vic-machine configure command

### LINKED ISSUE #4619: vic-machine reconfigure command
As vic-machine developer, I'd like to combine upgrade and reconfigure workflow, as upgrade is actually one special reconfiguration of VCH, to improve command usability and code maintainability. 

#4497 tracked the document for vic-machine reconfigure commands

Acceptance:
vic-machine upgrade works as before
vic-machine reconfigure works from end2end without any options included

### PR DESCRIPTION
**Pull request by emlin**
Use same process with upgrade.

Currently this command is still hidden, for two reasons:
- No any reconfigure option added (issues listed in #4641 )
- Convert data from guestinfo back to command option or create data structure is not done yet ( #4618 ).

Fixes #4619

### CODE REVIEW

File: cmd/vic-machine/configure/configure.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+	return configure
+}
+
+// Flags return all cli flags for upgrade
+func (c *Configure) Flags() []cli.Flag {
+	util := []cli.Flag{
+		cli.BoolFlag{
+			Name:        "force, f",
+			Usage:       "Force the configure",
```
Comment (anchal-agrawal) [inline cmd/vic-machine/configure/configure.go]: `Force the configure operation` sounds better.

File: cmd/vic-machine/configure/configure.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+	if !c.Data.Rollback {
+		err = executor.Configure(vch, vchConfig, vConfig)
+	} else {
+		err = executor.Rollback(vch, vchConfig, vConfig)
+	}
+
+	if err != nil {
+		// upgrade failed
+		executor.CollectDiagnosticLogs()
+		if err == nil {
```
Comment (anchal-agrawal) [inline cmd/vic-machine/configure/configure.go]: Would `err == nil` be possible here?

File: lib/install/data/data_test.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
```
Comment (anchal-agrawal) [inline lib/install/data/data_test.go]: 2017

File: lib/install/data/data.go [hunk 4]
```diff
@@ -134,12 +149,35 @@ type InstallerData struct {
...
+			MappedNetworksIPRanges: make(map[string][]ip.Range),
+			MappedNetworksDNS:      make(map[string][]net.IP),
+		},
+		Timeout: 3 * time.Minute,
 	}
 	return d
 }
+
+// CopyNonEmpty will shallow copy src value to override existing value if the value is set
+// This copy will take care of relationship between variables, that means if any variable in ContainerNetwork or NetworkConfig is not empty, the whole ContainerNetwork or NetworkConfig will be copied
```
Comment (anchal-agrawal) [inline lib/install/data/data.go]: It'd be nice to break this line into two for readability.

File: cmd/vic-machine/configure/configure.go [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+			Destination: &c.Timeout,
+		},
+		cli.BoolFlag{
+			Name:        "resetInProgressFlag",
+			Usage:       "Reset the UpdateInProgress flag. Warning: Do not reset this flag if another upgrade/configure process is running",
+			Destination: &c.ResetInProgressFlag,
+		},
+		cli.BoolFlag{
+			Name:        "rollback",
+			Usage:       "Roll back VCH configuration to before the previous upgrade/configure",
```
Comment (andrewtchin) [inline cmd/vic-machine/configure/configure.go]: Should this be "to before the current upgrade/configure"?

File: lib/install/management/configure.go [hunk 6]
```diff
@@ -240,7 +254,11 @@ func (d *Dispatcher) update(conf *config.VirtualContainerHostConfigSpec, setting
 		}
 	}
 
-	if err = d.reconfigVCH(conf, fmt.Sprintf("[%s] %s/%s", conf.ImageStores[0].Host, d.vmPathName, settings.ApplianceISO)); err != nil {
+	isoFile := settings.ApplianceISO
```
Comment (anchal-agrawal) [inline lib/install/management/configure.go]: Here, if `settings.ApplianceISO == ""`, `isoFile = ""` as well.

```go
        isoFile := ""
 	if settings.ApplianceISO != "" {
 		isoFile = fmt.Sprintf("[%s] %s/%s", conf.ImageStores[0].Host, d.vmPathName, settings.ApplianceISO)
 	}
```

The above snippet is easier to understand and has the same behavior as the original code.

File: lib/install/data/data_test.go [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+package data
+
+import (
+	"net"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestCopy(t *testing.T) {
```
Comment (anchal-agrawal) [inline lib/install/data/data_test.go]: We should rename this func to `TestCopyNonEmpty`. Also, can we add cases to check that empty fields are not copied?

File: cmd/vic-machine/configure/configure.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+			Destination: &c.Force,
+		},
+		cli.DurationFlag{
+			Name:        "timeout",
+			Value:       3 * time.Minute,
+			Usage:       "Time to wait for configure",
+			Destination: &c.Timeout,
+		},
+		cli.BoolFlag{
+			Name:        "resetInProgressFlag",
```
Comment (andrewtchin) [inline cmd/vic-machine/configure/configure.go]: Can this be changed to `reset-progress` or something since none of the other flags are camel case?

Comment (emlin) [inline cmd/vic-machine/configure/configure.go]: hmm.. that is used in upgrade command. Anyway, let me change both of them

File: cmd/vic-machine/configure/configure.go [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+	}
+	if err != nil {
+		log.Error("Failed to get Virtual Container Host configuration")
+		log.Error(err)
+		return errors.New("configure failed")
+	}
+
+	// TODO: Convert guestinfo *VirtualContainerHost back to *Data, decrypt secret data
+	oldData := &data.Data{}
+	// oldData, err := converter.GtoD(vchConfig)
```
Comment (anchal-agrawal) [inline cmd/vic-machine/configure/configure.go]: Unneeded comment?

Comment (emlin) [inline cmd/vic-machine/configure/configure.go]: This is the comment intentionally left here. Those code is not there yet, but I want to have them added here.

**Comment by anchal-agrawal** [comment]
@emlin, can we add a `log.Infof("")` on line 209 in [this file](https://github.com/vmware/vic/blob/master/lib/install/management/inspect.go#L209) so we have a blank INFO line between the IP and the `Unable to find valid client certs` message?

picture

This request came from George during the sprint demo and it'd be nice to make this change here since it's related and has zero impact.

**Pull request merged**